client/scripts/controllers/chain/substrate/staking.ts

```diff
@@ -226,6 +257,93 @@ class SubstrateStaking implements StorageModule {
+    return this._Chain.api.pipe(
+      switchMap((api: ApiRx) => combineLatest(
+        of(api),
+        api.derive.staking.validators(),
+        api.query.staking.currentEra(),
+        from(this._app.chainEvents.rewards()),
+        api.derive.staking.electedInfo()
+      )),
+      flatMap((
+        [api, { validators: currentSet }, era, rewards, electedInfo]:
+        [ApiRx, DeriveStakingValidators, EraIndex, IReward, DeriveStakingElected]
+      ) => {
+        const commission: ICommissionInfo = {};
+
+        electedInfo.info.forEach(({ accountId, validatorPrefs }) => {
+          const key = accountId.toString();
+          commission[key] = (validatorPrefs.commission.unwrap() || new BN(0)).toNumber() / 10_000_000;
+          return commission[key];
+        });
+        // Different runtimes call for different access to stakers: old vs. new
+        const stakersCall = (api.query.staking.stakers)
+          ? api.query.staking.stakers
+          : api.query.staking.erasStakers;
+        // Different staking functions call for different function arguments: old vs. new
+        const stakersCallArgs = (account) => (api.query.staking.stakers)
+          ? account
+          : [era.toString(), account];
+        return combineLatest(
+          stakersCall.multi(currentSet.map((elt) => stakersCallArgs(elt.toString()))),
+          of(currentSet),
+          of(rewards),
+          of(commission)
+        );
+      }),
+      auditTime(100),
+      map(([exposures, accounts, rewards, commissions ] :
+        [Exposure[], AccountId[], IReward, ICommissionInfo ]) => {
+        // coins.div don't support fraction points. multiply and divide the same number will give fraction points.
+        const n = 100000;
+        const validatorRewards: ICommissionInfo = {};
+        let accountsTotalStake = new BN(0);
+
+        accounts.forEach((account, index) => {
+          const key = account.toString();
+          const exposure = exposures[index];
+          let stakeShare = 0;
+          let othersStake: BN = new BN(0);
+
+          exposure.others.forEach((indv) => {
+            othersStake = othersStake.add(indv.value.toBn());
+          });
+          othersStake = othersStake.add(exposure.own.toBn());
+          if (othersStake.gt(new BN(0))) {
+            stakeShare = exposure.own.toBn().muln(n).div(othersStake).toNumber() / n;
+          }
+
+          const totalStake = exposure?.total.toBn() || new BN(0);
+          let totalReward = rewards.validators[key] || 0;
+
+          // if rewards have rewards for chain id, split with every validator equally.
+          if (rewards.validators[this._app.chain.id])
+            totalReward += rewards.validators[this._app.chain.id] / accounts.length;
+
+          accountsTotalStake = accountsTotalStake.add(totalStake);
+
+          const commissionPercent: number = (commissions[key] || 0) / 100;
+
+          let rewardEarn = totalReward * commissionPercent;
```

Comment:
The math here should use `BN` struct since it deals with large numbers. Comments would also be appreciated.

Reply:
Number too large error on converting to BN on `totalReward`. So I haven't updated.